micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/GrpcTagProvider.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2021 VMware, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.micrometer.core.instrument.binder.grpc;
+
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.micrometer.core.instrument.Tag;
+
+public interface GrpcTagProvider {
```

Comment:
This is not solving what was described in https://github.com/micrometer-metrics/micrometer/issues/2622.. There should be a way to add additional tags for each call.. This should work for both client and server side..

Reply:
Correct. See my comment here: https://github.com/micrometer-metrics/micrometer/pull/2810#issuecomment-944392883

The current implementation computes and caches all the tags for a `MethodDescriptor`. Adding the ability to tag based on `Metadata` as well would pretty much remove that optimization, resulting in new tags created per call instead of the current cache based solely on `MethodDescriptor`

I was awaiting feedback on that decision.

Reply:
I actually just thought of something. Instead of a single tagprovider, perhaps there could be 2 different interfaces. One w/ just MethodDescriptor and is cached, the other with both and is computer per call. It's been a while since I wrote this, I'll take a peek and see if we could accomplish both.